### ISSUE TITLE
CSA --Wanalyzer-null-dereference false negative with *p = 42

### ISSUE THREAD

**Issue by Geoffrey1014** [llvm/llvm-project#59735]
I got a false negative error when compiling the following program with clang(trunk) `-Xanalyzer -analyzer-output=text --analyze -Xanalyzer -analyzer-checker=debug.ExprInspection` in https://godbolt.org/z/vE9E4j96x. 

In this case, the evaluation result for `0 == p` on line 21 is TRUE, so clearly clang static analyzer has known that the pointer p is a null pointer there, but then the analyzer does not generate an NPD warning for the `*p = 42` statement on line 22. I think it's a little odd. Here is the result of the program, please take a look, thank you.
```c
#include "stdio.h"
void clang_analyzer_eval(int);
void clang_analyzer_warnIfReached();
int *f(int *);

int *f(int *p)
{
    p = (int *)0;
    return p;
}

int main()
{
    int a = 42;
    int *p = &a;
    clang_analyzer_eval(0 == p);
    p = f(p);
    clang_analyzer_eval(0 == p);
    if (p == (int *)0)
    {
        clang_analyzer_eval(0 == p);
        *p = 42;
    }
}
```
```
<source>:16:5: warning: FALSE [debug.ExprInspection]
    clang_analyzer_eval(0 == p);
    ^~~~~~~~~~~~~~~~~~~~~~~~~~~
<source>:16:5: note: FALSE
    clang_analyzer_eval(0 == p);
    ^~~~~~~~~~~~~~~~~~~~~~~~~~~
<source>:18:5: warning: TRUE [debug.ExprInspection]
    clang_analyzer_eval(0 == p);
    ^~~~~~~~~~~~~~~~~~~~~~~~~~~
<source>:18:5: note: TRUE
    clang_analyzer_eval(0 == p);
    ^~~~~~~~~~~~~~~~~~~~~~~~~~~
<source>:21:9: warning: TRUE [debug.ExprInspection]
        clang_analyzer_eval(0 == p);
        ^~~~~~~~~~~~~~~~~~~~~~~~~~~
<source>:19:9: note: 'p' is equal to null
    if (p == (int *)0)
        ^
<source>:19:5: note: Taking true branch
    if (p == (int *)0)
    ^
<source>:21:9: note: TRUE
        clang_analyzer_eval(0 == p);
        ^~~~~~~~~~~~~~~~~~~~~~~~~~~
```

**Comment by steakhal** [comment]
This is due to the inline defensive checks supression heuristic. At some point we could rework the heuristic to be a little smarter. @haoNoQ must know some cross refs, or the umbrella ticket for this.
